### PR TITLE
Added PerformanceModule annotation

### PR DESCRIPTION
**Pull request by gupele**

### CODE REVIEW

File: core/src/main/java/com/microsoft/applicationinsights/internal/annotation/PerformanceModule.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ */
+
+import java.lang.annotation.Target;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Target({ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+public @interface PerformanceModule {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/annotation/PerformanceModule.java]: I wonder what we'll do with the other annotations for telemetry modules/initializers - we already have TelemetryModule / TelemetryInitializer classes... WDTH?

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 2]
```diff
@@ -128,6 +133,51 @@ private void setInternalLogger(ConfigFileParser parser, TelemetryConfiguration c
...
+            public Class<? extends Annotation>[] annotations() {
+                return new Class[]{PerformanceModule.class};
+            }
+
+            @Override
+            public void reportTypeAnnotation(Class<? extends Annotation> annotation, String className) {
+                performanceModuleNames.add(className);
+            }
+        };
+        final AnnotationDetector cf = new AnnotationDetector(reporter);
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: what is cf?

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 3]
```diff
@@ -128,6 +133,51 @@ private void setInternalLogger(ConfigFileParser parser, TelemetryConfiguration c
...
+            @SuppressWarnings("unchecked")
+            public Class<? extends Annotation>[] annotations() {
+                return new Class[]{PerformanceModule.class};
+            }
+
+            @Override
+            public void reportTypeAnnotation(Class<? extends Annotation> annotation, String className) {
+                performanceModuleNames.add(className);
+            }
+        };
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: please move out do a dedicated class

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 4]
```diff
@@ -128,6 +133,51 @@ private void setInternalLogger(ConfigFileParser parser, TelemetryConfiguration c
...
+        final AnnotationDetector cf = new AnnotationDetector(reporter);
+        try {
+            cf.detect("com.microsoft.applicationinsights");
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: I think it is better to encapsulate this entire logic of the TypeReporter & AnnotationDetector.
A class with a simple 'detect' API for a given annotation type. The code will be cleaner.

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 5]
```diff
@@ -128,6 +133,51 @@ private void setInternalLogger(ConfigFileParser parser, TelemetryConfiguration c
...
+            for (String performanceModuleName : performanceModuleNames) {
+                TelemetryModule module = null;
+                try {
+                    module = createInstance(performanceModuleName, TelemetryModule.class);
+                    if (module != null) {
+                        modules.add(module);
+                    } else {
+                        InternalLogger.INSTANCE.error("Failed to create performance module: '%s'", performanceModuleName);
+                    }
+                } catch (Exception e) {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: This is redundant - createInstance can't throw any exception.

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java [hunk 6]
```diff
@@ -128,6 +133,51 @@ private void setInternalLogger(ConfigFileParser parser, TelemetryConfiguration c
         InternalLogger.INSTANCE.initialize(loggerOutput, loggerData.items);
     }
 
+    private List<TelemetryModule> setPerformanceCounters(ConfigFileParser parser) {
```
Comment (Yonisha) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: getPerformanceCounters (the method doesn't set anything :))

**Comment by Yonisha** [comment]
signed off

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java (hunk 6 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: right, I will update the name thanks

File: core/src/main/java/com/microsoft/applicationinsights/internal/annotation/PerformanceModule.java (hunk 1 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/annotation/PerformanceModule.java]: It is a good question, I thought that we should go with self explanatory names, like PerformanceModule, its location under package 'annotation' should help distinguish, I think we should try to make sure that it is clear to our users

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java (hunk 4 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: I think you are right, I will that on the next pull request

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java (hunk 3 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: I agree

File: core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java (hunk 2 above)
Comment (gupele) [inline core/src/main/java/com/microsoft/applicationinsights/internal/config/TelemetryConfigurationFactory.java]: I don't know :) will change

**Pull request merged**